### PR TITLE
Adding additional test for CL && Correcting UL element

### PR DESCRIPTION
**Pull request by loumalouomega**
Making happy @RiccardoRossi 

@pooyan-dadvand Can we add the tag test?

### CODE REVIEW

**Comment by pooyan-dadvand** [comment]
Added!

**Comment by loumalouomega** [comment]
Thanks

**Comment by loumalouomega** [comment]
@RiccardoRossi I included the UL correction

**Comment by loumalouomega** [comment]
@RiccardoRossi I included the UL correction

**Comment by loumalouomega** [comment]
@RiccardoRossi Ping

**Comment by loumalouomega** [comment]
Ping

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp]: ok, however you'll have to reimplement it for UL if you want to be able to remesh

File: applications/StructuralMechanicsApplication/tests/constitutive_law_test.py
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/tests/constitutive_law_test.py:80]: why are u doing this?

File: applications/StructuralMechanicsApplication/tests/constitutive_law_test.py
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/tests/constitutive_law_test.py:167]: wouldn't it be better to set it directly here??

File: applications/StructuralMechanicsApplication/tests/constitutive_law_test.py
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/tests/constitutive_law_test.py:234]: same comment

File: applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:79]: ok with the change, but may i ask you why this decision?

File: kratos/solving_strategies/strategies/residualbased_linear_strategy.h
Comment (RiccardoRossi) [inline kratos/solving_strategies/strategies/residualbased_linear_strategy.h:479]: this change touches one of the important classes. As such it must be discussed on its own (I for one, do not love the name GetDirectSystemMatrix) 

The correct way would have been to make a PR only with those changes. 

for the moment let's leave it here, but i'll block the PR until we agree this is correct.
@pooyan-dadvand @jcotela comments on this? also @AndreasWinterstein, i think you also needed this capability some time back...

I have 2:
1 - I would call this GetSystemMatrix
2 - I am not clear of what happens when we are doing 
     
     A = *mpA. I think (but i may be wrong) that this triggers a copy and not a reference.

if this is so i would have implemented it as

    
     const TSystemMatrixType& GetSystemMatrix()
     {
         return *mpA;

**Review by RiccardoRossi — COMMENTED** [review]
overall the PR is fine. I am waiting for approval because of the functions
 
GetDirectSystemMatrix

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp]: Why, because the initial coordinates will be lost?

File: applications/StructuralMechanicsApplication/tests/constitutive_law_test.py
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/tests/constitutive_law_test.py:80]: What do you mean?

File: applications/StructuralMechanicsApplication/tests/constitutive_law_test.py
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/tests/constitutive_law_test.py:167]: I think is more modular to do it this way

File: applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:79]: Because it was giving me problems (the jacobian appeared as inverted in first iteration)

File: kratos/solving_strategies/strategies/residualbased_linear_strategy.h
Comment (loumalouomega) [reply kratos/solving_strategies/strategies/residualbased_linear_strategy.h:479]: The way you suggest is the way is already implemented, but it was giving me problems to compile in python because the type of matrix was condensed and I try with the condensed matrix, but that gave me errors. So that why i created this method

File: applications/StructuralMechanicsApplication/tests/constitutive_law_test.py
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/tests/constitutive_law_test.py:80]: why using a dictionary i mean.

File: applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:79]: mmmmm

can u check again after the latest corrections?

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:79]: I can

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:79]: Nope, doesn't work

File: applications/StructuralMechanicsApplication/tests/constitutive_law_test.py
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/tests/constitutive_law_test.py:80]: Because there are many options and dictionaries are useful

**Comment by loumalouomega** [comment]
PING

File: applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:79]: Can you check if it works with the TL? If it doesn't keep it as it is

File: kratos/solving_strategies/strategies/residualbased_linear_strategy.h
Comment (RiccardoRossi) [reply kratos/solving_strategies/strategies/residualbased_linear_strategy.h:479]: Please remove the new method. We need instead to fix the old method if it is not working

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp]: Just read the comment... If you remedy this assumption is not valid (it is where UL and TL are no longer identical)

File: applications/StructuralMechanicsApplication/custom_elements/total_lagrangian.cpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/total_lagrangian.cpp]: This is very slow. You should not pass the coordinates, only the is of the pointer! (I see the t it is not a change you did, still I see the error and it should be corrected)

File: applications/StructuralMechanicsApplication/custom_elements/total_lagrangian.cpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/total_lagrangian.cpp]: Same

File: applications/StructuralMechanicsApplication/tests/pendulus_test/pendulus_TL_test.mdpa
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/tests/pendulus_test/pendulus_TL_test.mdpa:1]: Why so many elements? Can t you make the test smaller?

File: applications/StructuralMechanicsApplication/tests/pendulus_test/pendulus_UL_test.mdpa
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/tests/pendulus_test/pendulus_UL_test.mdpa:1]: Make it smaller. Also ideally create the test without mode, to the style of the patch test

File: applications/StructuralMechanicsApplication/tests/constitutive_law_test.py
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/tests/constitutive_law_test.py:80]: The CO already has an internal dictionary of options. I Don t understand why you would like to replicate that

File: applications/StructuralMechanicsApplication/tests/constitutive_law_test.py
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/tests/constitutive_law_test.py:167]: Honestly I disagree... @pooyan-dadvand what do you think?

File: kratos/solving_strategies/strategies/residualbased_linear_strategy.h
Comment (loumalouomega) [reply kratos/solving_strategies/strategies/residualbased_linear_strategy.h:479]: OK, I need help with the boost::python then

File: applications/StructuralMechanicsApplication/custom_elements/total_lagrangian.cpp
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_elements/total_lagrangian.cpp]: What?

File: applications/StructuralMechanicsApplication/custom_elements/total_lagrangian.cpp
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_elements/total_lagrangian.cpp]: The this-> is slow?

File: applications/StructuralMechanicsApplication/tests/pendulus_test/pendulus_TL_test.mdpa
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/tests/pendulus_test/pendulus_TL_test.mdpa:1]: For the pendulus, is already quite small

File: applications/StructuralMechanicsApplication/tests/pendulus_test/pendulus_UL_test.mdpa
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/tests/pendulus_test/pendulus_UL_test.mdpa:1]: mdpa?

File: applications/StructuralMechanicsApplication/custom_elements/total_lagrangian.cpp
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/custom_elements/total_lagrangian.cpp]: No, what is slow is the call

ShapeFunctionsValues(vector, coordinates)

You should use instead the one that returns at once the one on all the integration points, smug like

ShapeFunctionsValues(matrix, integrstiontype)

Or eventually

ShapeFunctionsValues(vector, index)

File: applications/StructuralMechanicsApplication/tests/pendulus_test/pendulus_TL_test.mdpa
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/tests/pendulus_test/pendulus_TL_test.mdpa:1]: Well...you could use 5-6 elements instead of 50...

File: applications/StructuralMechanicsApplication/tests/pendulus_test/pendulus_UL_test.mdpa
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/tests/pendulus_test/pendulus_UL_test.mdpa:1]: Yes mdpa

File: applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:79]: It works

File: applications/StructuralMechanicsApplication/tests/constitutive_law_test.py
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/tests/constitutive_law_test.py:80]: Because this way I now the options I have and what I can set

File: applications/StructuralMechanicsApplication/tests/pendulus_test/pendulus_UL_test.mdpa
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/tests/pendulus_test/pendulus_UL_test.mdpa:1]: I prefer this way, but if this make you happy...

File: applications/StructuralMechanicsApplication/tests/constitutive_law_test.py
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/tests/constitutive_law_test.py:80]: i find this a rather weak argument. The list of options of the CL is documented in the wiki and should not change.

i really don't see how 

     cl.Set(USE_ELEMENT_PROVIDED_STRAIN, False) 

is worst than

     dict_options = {'USE_ELEMENT_PROVIDED_STRAIN': False, 

still i will accept this to end a sterile discussion

File: applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:79]: ... if it works with the TL and not with the UL, it means that the UL still has a problem. no?

File: kratos/solving_strategies/strategies/residualbased_linear_strategy.h
Comment (RiccardoRossi) [reply kratos/solving_strategies/strategies/residualbased_linear_strategy.h:479]: tomorrow when you are here we take a look at this. prepara a small test to verify how the existing implementation fails

**Review by RiccardoRossi — COMMENTED** [review]
to me this is almost ready to be accepted.

the two things missing are
1 - removing the method you added from the ResidualBasedLinearStrategy
2 - Discover why the UL fails while the TL works

File: applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:79]: The thing is that the deformation impossed previously inverted one element, this in TL is not a problem because the integration is done in the reference confguration, but not for the Ul. I did many tests, the TL and UL give exactly the same solution

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:79]: if you use an hyperelastic law, with linear elastic that's not true (except for very small deformations)

File: kratos/solving_strategies/strategies/residualbased_linear_strategy.h
Comment (loumalouomega) [reply kratos/solving_strategies/strategies/residualbased_linear_strategy.h:479]: The current test can do it, it just one line

**Comment by loumalouomega** [comment]
With the configuration you set the element is inverted and the problem can not be solved in the deformed configuration, look:
imagen

**Comment by loumalouomega** [comment]
@RiccardoRossi , can we push this?

File: kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h
Comment (RiccardoRossi) [inline kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h:827]: to be removed

File: kratos/solving_strategies/strategies/residualbased_linear_strategy.h
Comment (RiccardoRossi) [inline kratos/solving_strategies/strategies/residualbased_linear_strategy.h:480]: remove this function. There is already the function

GetSystemMatrix

**Comment by loumalouomega** [comment]
Done @RiccardoRossi

File: applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py]: the method GetDirectSystemMatrix is still there???

**Review by RiccardoRossi — APPROVED** [review]

**Pull request merged**
